### ISSUE TITLE
Adding Large Arrays is Extremely Slow

### ISSUE THREAD

**Issue by mattcbro** [JuliaDebug/JuliaInterpreter.jl#297]
I initially thought the problem was the dictionary lookup,  but it turns out that adding large arrays together is incredibly slow in the interpreter,  which really doesn't make any sense.  Shouldn't this drop down into a call into a BLAS library,  or should the line for adding the arrays get compiled into efficient code?

````julia
""" extract the raw data from dictionary """
function valdata(val)
    # next statement takes an infinite amount of time to step through
    xraw = val["rdat"] + 1.0im .* val["idat"]
    return(xraw)
end

ar = randn(3000,4000)
ai = randn(3000,4000)
vdat = Dict()
vdat["rdat"] = ar
vdat["idat"] = ai
vdat["num"] = pi
vdat["text"] = " This is text"

xcomp = Juno.@enter valdata(vdat)
````

**Comment by fredrikekre** [comment]
>  Shouldn't this drop down into a call into a BLAS library

Array addition uses julia-implemented broadcasting.

**Comment by mattcbro** [comment]
Nevertheless I do not understand why the interpreter performs this so slowly.  I'm guessing that it interprets the code all the way down to a bunch of interpreted for loops,  so the granularity is too fine.  It should simply compile each line into LLVM  bit code,  which could then be interpreted efficiently.  Hopefully things will improve,  but the speed makes debugging not particularly useful for me.

**Comment by KristofferC** [comment]
> I'm guessing that it interprets the code all the way down to a bunch of interpreted for loops, so the granularity is too fine.

Everything is interpreted down to the intrinsics.

> It should simply compile each line into LLVM bit code, which could then be interpreted efficiently.

This sounds like gobbledygook. Perhaps a short proof of concept would illustrate your idea better.

**Comment by StefanKarpinski** [comment]
I don’t think there’s anything specifically actionable here. It’s known that the interpreter-only debugger is slow and there’s a number of other discussions of possible approaches to better performance.

**Comment by mattcbro** [comment]
gobbledygook?  How so.  If you translate a line of julia code to llvm bit code,  it can be just in time compiled to machine code and executed.  Isn't that what the julia compiler essentially does with blocks of julia code when it is compiled/run?  
https://llvm.org/docs/CommandGuide/lli.html


The machine code should execute more efficiently than interpreting the code directly or translating it to some intermediate representation that apparently takes forever to interpret.  Efficiency is gained by either calling BLAS libraries or Julia's internal libraries for doing linear algebra.

I really don't see why this should be inefficient and indeed I've written my own front end stack machine for doing linear algebra operations that was very fast simply because it called into my systems BLAS libraries.  We are talking about simple array addition here!

Since Julia is a language noted for it's speed when doing math heavy compute tasks,  it makes sense to try to speed up the performance of the interpreter when debugging Julia code.

**Comment by KristofferC** [comment]
The problem is being able to inspect the full state of the whole stack and be able to breakpoint at arbitrary positions.

> it makes sense to try to speed up the performance of the interpreter when debugging Julia code.

Improving the performance does indeed make sense. We graciously accept any help you can give in terms of PRs that improves performance while keeping the debugging info intact.

**Comment by mattcbro** [comment]
I understand,  I had the same problem,  but a much simpler target language.   I saved my variables in a giant hash table and every line of code becomes a compiled function call.  I also used a stack in main memory to hold temporary variables.  I was constrained because I used C as an intermediary language.  No doubt you have more flexibility with LLVM. 

 You might ask yourself how does gdb work?  Apparently these debuggers use hardware specific traps.  There are flags on the x86 and ARM processors that puts the machine into single step mode,  and can query the instruction and register state of the processor.  The breakpoint hack inserts an invalid instruction at the break point which causes an exception that the debugger kindly handles,  whereupon it can execute the correct instruction if so desired.

The point of this is that the debugger can use the compiled code directly,  thereby not losing performance.  No doubt it's tricky finding the points in the binary corresponding to your instruction,  but that mapping is created when you add debugging symbols in the compile process.

https://blog.0x972.info/?d=2014/11/13/10/40/50-how-does-a-debugger-work

I'm sure you've seen all this before,  but maybe it will spark some ideas.

**Comment by StefanKarpinski** [comment]
@mattcbro, are you at all familiar with the history of debuggers for Julia? Gallium, etc? Your comments would suggest that you missed that whole bit of context. Let’s just say that linking to a “how do debuggers work” page not giving anyone here any new information.

Compiling each line of Julia code to LLVM doesn’t really make sense because lines of code are not a reasonable unit of compilation. At that point you might as well interpret which is what we’re doing. Especially if you compile such that everything can be fully inspected in a debugger.

Yes, calling BLAS would be fast, but so would just calling Julia’s compiled implementation of this array op. But then it won’t stop at breakpoints or do any of the other things that make a debugger what it is. We could call compiled versions of operations that don’t need any debugger intervention—but knowing when that is or isn’t the case is precisely the problem.

**Comment by mattcbro** [comment]
@StefanKarpinski I am not fully cognizant of the history of Gallium outside of inspecting some of the source files at one point to try to understand how I could do some introspection on Julia's local variables,  if at all.

That brief look into the issue showed that the approach taken seemed very complex,  requiring a very deep dive into parsing and compiling Julia code.  That approach differs strongly from my understanding of how a debugger like gdb works.  I'm guessing that's where you took a hit in performance.  You may have gained some additional flexibility,  with many useful applications,  but you lost in performance.

That's my very high level understanding,  and I'm sure it's not the complete picture,  nor would I pretend to tell you how to solve this problem,  since the team has been at it for a while and obviously has had to sort through a lot of issues.

For the purpose of a simple debugger however,  a " line" of Julia code is precisely the unit of compilation you need LOL.  However since LLVM has source level debugging support I'm curious if that was exploited since it sidesteps that problem. The code I looked at in Rebugger a while ago, as an example, does a lot of Julia level code analysis instead of using the built in tools available in LLVM.  It looks like LLVM does a lot of work to maintain the mapping between the source,  the AST and generated LLVM code and would be very useful for this task.

I think this is where my ignorance of the full scope of the problem comes in.  I'm guessing that there is a lot more going on under the hood than simply executing LLVM bit code.  In the source I noticed weird things like lisp code and so forth that was definitely making my eyes blurry.  Perhaps this forces the full parsing approach rather than simply tagging the resulting executable with debugging symbol information.

**Comment by pfitzseb** [comment]
If all you want is to run through/step into some code and don't care about breakpoints at all, then take a look at the `Run in Compiled Mode` toggle in Juno or the `C: toggle compiled mode` option in Debugger.jl -- with those you can run your code fairly quickly:
```
julia> @time Juno.@run valdata(vdat)
  0.188299 seconds (297 allocations: 366.226 MiB, 18.65% gc time)
```

**Comment by StefanKarpinski** [comment]
Gallium is a full gdb-style debugger. That approach was fast but too unreliable to be useful. In short, we’ve already tried all the things you’re suggesting and it did not go well for many reasons. Compilers do not emit good debug information, the info they do generate is not good enough to debug s language like Julia, and the various other issues. The fact that gdb works is an illusion created by co-evolution of just what debug info is not completely broken and the limited kinds of things you can look at in gdb. 

This new debugger is the result of learning from that lesson and just implementing a very reliable but slow interpreting debugger. It is slow but it works and is useful. The parts of Gallium that you’re looking at are actually the origin of this debugger and we’re an attempt to patch up the problems with the gdb-like approach by mixing it with some interpretation. That was only added after it was found that a pure gdb-like approach was unworkable.

The bottom line is that everyone here knows how “real” debuggers work. The problem you’re encountering is known and was anticipated before this was even built. There is no “why don’t you just” solution to this. If you want to help make this debugger better, great! If not, that’s ok too. But if you do want to help, please approach it from the stance that the people working on this know what they are doing.

**Comment by mattcbro** [comment]
I'm not suggesting anything,  nor I am I trying to tell you how to solve this problem, nor am I intending to question anyone's expertise.  I am just trying to understand why one is forced to take hit in array computation performance in the interpreter,  or why the granularity of the compilation/interpolation prevents calls to BLAS etc.

As an aside,  how difficult would it be to implement something like a matlab keyboard command?  There is support now for getting local variable names and values right?  This would be a good work around for my application I think,  especially if I could inspect local and global variables even if only read only.

**Comment by KristofferC** [comment]
> why the granularity of the compilation/interpolation prevents calls to BLAS etc.

This has nothing to do with the debugger. Julia just doesn't call BLAS for array addition so of course interpreting that code will not call BLAS either.

**Comment by pfitzseb** [comment]
> I am just trying to understand why one is forced to take hit in array computation performance in the interpreter, or why the granularity of the compilation/interpolation prevents calls to BLAS etc.

Because all Julia code is completly interpreted in the debugger, and that needs to happen so you can set breakpoints in every Julia function. It doesn't at all prevent calls to BLAS:
```
julia> A, B = rand(3000,3000), rand(3000,3000);

julia> @time Juno.@run A*B;
  0.512220 seconds (20.55 k allocations: 69.664 MiB, 2.65% gc time)
```
Broadcasting on the other hand is fully implemented in Julia, so `A .+ B` would mean interpreting 9e6+ function calls.

> As an aside, how difficult would it be to implement something like a matlab keyboard command? There is support now for getting local variable names and values right? This would be a good work around for my application I think, especially if I could inspect local and global variables even if only read only.

[Fairly easy](https://gist.github.com/pfitzseb/2cd6758eec7ad52a5a4b4421400f9602), if all you want is readonly access to variables:
```
julia> function f(x)
         y = x
         z = x + y
         @keyboard
         x*y*z
       end
f (generic function with 1 method)

julia> f(3)
debug> x
3

debug> y
3

debug> z
6

debug> 
54
```

**Comment by mattcbro** [comment]
@pfitzseb That little keyboard macro is very very useful.  Thanks a bunch.  It nicely sidesteps my problem.  Will this be part of the Interpreter or Debugger package any time soon?

**Comment by KristofferC** [comment]
We could add something like it to e.g. Debugger.jl, yes.

**Comment by StefanKarpinski** [comment]
Maybe call it something besides `@keyboard`?

**Comment by boubalos** [comment]
> If all you want is to run through/step into some code and don't care about breakpoints at all, then take a look at the `Run in Compiled Mode` toggle in Juno or the `C: toggle compiled mode` option in Debugger.jl -- with those you can run your code fairly quickly:
> 
> ```
> julia> @time Juno.@run valdata(vdat)
>   0.188299 seconds (297 allocations: 366.226 MiB, 18.65% gc time)
> ```

This is awesome!
For expresion such "C=A+B", actually I do not need to probe into the details of A+B, I just what the result of A+B.  so that's definitely solved my problem!
